### PR TITLE
[hardware] feat: add Intel XPU device detection

### PR DESCRIPTION
**Pull request by kahlun**
Add Intel XPU as a supported device type alongside CUDA and NPU.
One file changed, pure detection only, zero behavioral changes to existing CUDA/NPU paths.

Changes in verl/utils/device.py:
- is_torch_xpu_available(check_device=True) + is_xpu_available module-level flag
- get_device_name() → "xpu"
- get_nccl_backend() → "xccl"
- get_resource_name() → "XPU" (Ray placement)
- get_visible_devices_keyword() → "ONEAPI_DEVICE_SELECTOR"
- auto_set_device() → corrects config.trainer.device to "xpu" when defaulting to "cuda"

Every change follows the existing NPU elif pattern exactly.

**Hardware:** Intel Arc Pro B60 (24 GB, PCIe), 4-card system
**PyTorch:** 2.10+xpu / 2.11+xpu

### What does this PR do?

Add Intel XPU as a supported device type in `verl/utils/device.py`, alongside the existing CUDA and Ascend NPU support.

This is the **foundation PR** for XPU enablement — 1 file, pure detection, zero behavioral changes to existing paths. All other XPU PRs (FSDP workers, VLM attention, vLLM rollout, XCCL workarounds, TorchTitan) depend on this.

No similar PRs exist: [XPU search](https://github.com/verl-project/verl/pulls?q=is%3Apr+XPU) | [Intel search](https://github.com/verl-project/verl/pulls?q=is%3Apr+Intel)

**Reference:** Follows the same enablement pattern as pytorch/torchtitan [PR #672](https://github.com/pytorch/torchtitan/pull/672) — device abstraction first, incremental features in follow-up PRs.

**Follow-up PRs planned (on [fork](https://github.com/kahlun/verl/pulls)):**

| PR | What | Files |
|----|------|-------|
| A1 | XCCL workarounds (ReduceOp.AVG/MAX) | 2 |
| A2 | Ray resource mapping | 2 |
| B  | FSDP workers + SDPA attention default | 9 |
| C  | VLM attention (SDPA replacement for flash_attn) | 5 |
| D  | vLLM rollout fixes | 5 |
| E1 | TorchTitan XPU | 2 |
| E2 | TorchTitan Pipeline Parallel | 1 |
| F  | Fused kernel guards | 2 |

### Checklist Before Starting

- [x] Search for similar PRs:
  - https://github.com/verl-project/verl/pulls?q=is%3Apr+XPU — 0 results
  - https://github.com/verl-project/verl/pulls?q=is%3Apr+xpu — 0 results
  - https://github.com/verl-project/verl/pulls?q=is%3Apr+Intel+device — 0 results
- [x] PR title: `[hardware] feat: add Intel XPU device detection`

### Test

**Hardware:** Intel Arc Pro B60 (24 GB, PCIe) × 4, PyTorch 2.10+xpu

**E2E — GRPO 1-GPU, 20 steps (vLLM rollout → FSDP train loop):**
- Training backend: FSDP
- Rollout backend: vLLM
- Model: Qwen2.5-0.5B-Instruct

```
step= 1 | loss=1.34e-06 | grad_norm=1.41e-04 | gen=16.3s | update_actor=6.8s
step=10 | loss=1.00e-06 | grad_norm=1.22e-04 | gen=8.8s  | update_actor=4.8s
step=20 | loss=8.63e-07 | grad_norm=7.62e-05 | gen=8.9s  | update_actor=4.9s
```
Loss decreasing, training stable over 20 steps ✓

**SFT — 1 GPU, 5 steps (training only, no rollout):**
- Model: Qwen2.5-0.5B-Instruct
- Checkpoint saved at global_step_5 ✓

**Note:** This PR (device detection only, 1 file) has zero impact on training or rollout code paths. The test results above validate the overall XPU stack which the follow-up PRs implement.

### API and Usage Example

No API changes. Existing code continues to work unchanged on CUDA/NPU.

```python
# On XPU hardware — new values returned:
from verl.utils.device import get_device_name, get_nccl_backend, is_xpu_available
print(is_xpu_available)      # True
print(get_device_name())     # "xpu"
print(get_nccl_backend())    # "xccl"

# On CUDA/NPU — zero behavioral change
```

### Design & Code Changes

Single file: `verl/utils/device.py` (+49/-6)

Every change is an `elif is_xpu_available` branch added after the existing `elif is_npu_available` — exact same pattern.

### Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] pre-commit checks passed: `pre-commit run --files verl/utils/device.py` — all 12 checks passed (ruff, mypy, license, etc.)
- [x] No documentation changes needed — detection functions are internal utilities.
- [x] Unit tests: XPU detection requires XPU hardware; validated manually on Intel Arc Pro B60. CI cannot run on XPU hardware.
- [ ] CI request: will post in #ci-request on verl Slack once reviewer approves direction.
- [x] Not related to recipe submodule.

### CODE REVIEW

File: verl/utils/device.py [hunk 1]
```diff
@@ -43,16 +43,31 @@ def is_torch_npu_available(check_device=True) -> bool:
         return False
 
 
+def is_torch_xpu_available() -> bool:
+    """Check the availability of XPU"""
+    try:
+        return torch.xpu.is_available()
+    except (ImportError, AttributeError):
+        return False
```
[the remark this answers is not in the document]
Comment (kahlun) [reply verl/utils/device.py:51]: Good catch — accepted. Added `check_device` parameter to `is_torch_xpu_available()` matching the NPU pattern. Updated in latest push.

File: verl/utils/device.py [hunk 2]
```diff
@@ -43,16 +43,31 @@ def is_torch_npu_available(check_device=True) -> bool:
...
+        ray resource name string, either "GPU", "NPU", or "xpu".
     """
-    return "GPU" if is_cuda_available else "NPU"
+    if is_cuda_available:
+        return "GPU"
+    elif is_npu_available:
+        return "NPU"
+    elif is_xpu_available:
+        return "xpu"
+    return "GPU"
```
[the remark this answers is not in the document]
Comment (kahlun) [reply verl/utils/device.py:82]: Agreed — already fixed in previous push. `"XPU"` uppercase and fallback `return "NPU"` to preserve original behavior.

File: verl/utils/device.py [hunk 3]
```diff
@@ -65,7 +80,13 @@ def get_visible_devices_keyword() -> str:
...
             'ASCEND_RT_VISIBLE_DEVICES' otherwise.
     """
-    return "CUDA_VISIBLE_DEVICES" if not is_torch_npu_available(check_device=False) else "ASCEND_RT_VISIBLE_DEVICES"
+    if is_cuda_available:
+        return "CUDA_VISIBLE_DEVICES"
+    elif is_npu_available:
+        return "ASCEND_RT_VISIBLE_DEVICES"
+    elif is_xpu_available:
+        return "ONEAPI_DEVICE_SELECTOR"
+    return "CUDA_VISIBLE_DEVICES"
```
[the remark this answers is not in the document]
Comment (kahlun) [reply verl/utils/device.py:101]: Valid point — fixed. Restored original NPU behavior by using `is_torch_npu_available(check_device=False)` (package check, not device check). XPU added the same way with `is_torch_xpu_available(check_device=False)`. Updated in latest push.

**Comment by wuxibin89** [comment]
Have you run e2e training on XPU, which training and rollout backend do you use?

File: verl/utils/device.py [hunk 4]
```diff
@@ -124,6 +147,8 @@ def get_nccl_backend() -> str:
     """
     if is_npu_available:
         return "hccl"
+    elif is_xpu_available:
+        return "xccl"
```
[the remark this answers is not in the document]
Comment (kahlun) [reply verl/utils/device.py:165]: Verified on PyTorch 2.10.0+xpu:

```python
>>> torch.distributed.is_backend_available("xccl")
True
>>> torch.distributed.is_backend_available("ccl")
False
```

`"xccl"` is the correct native backend name. The old `"ccl"` was from the now-deprecated external `intel/torch-ccl` package ([archived Dec 2025](https://github.com/intel/torch-ccl)).

In PyTorch core: `Backend.XCCL = "xccl"` (added in [pytorch/pytorch#141856](https://github.com/pytorch/pytorch/pull/141856), Dec 2024). The implementation is `ProcessGroupXCCL`. There is no `ProcessGroupCCL` or `Backend.CCL` in current PyTorch.

Official PyTorch distributed docs list four built-in backends: `gloo`, `mpi`, `nccl`, `xccl`.

**Comment by kahlun** [comment]
> Have you run e2e training on XPU, which training and rollout backend do you use?

Yes, e2e training runs on 4× Intel Arc Pro B60 (PyTorch 2.10+xpu):

Training: FSDP backend — SFT/GRPO/PPO validated (1-GPU and 2-GPU)
Rollout: vLLM
This PR is just the detection layer — the first in a series. Follow-up PRs for FSDP workers, attention, vLLM rollout, and XCCL workarounds are ready on [our fork](https://github.com/kahlun/verl/pulls) and will be submitted after this one merges.

**Comment by kahlun** [comment]
To clarify my earlier reply:

**Training backend:** FSDP
**Rollout backend:** vLLM

GRPO 1-GPU, 20 steps — full e2e loop (vLLM rollout → reward → FSDP train → repeat):
- Model: Qwen2.5-0.5B-Instruct, Dataset: GSM8K, Hardware: Intel Arc Pro B60

```
step= 1 | loss=1.34e-06 | grad_norm=1.41e-04 | rollout=16.3s | train=6.8s
step= 5 | loss=1.30e-06 | grad_norm=1.35e-04 | rollout=9.1s  | train=4.8s
step=10 | loss=1.00e-06 | grad_norm=1.22e-04 | rollout=8.8s  | train=4.8s
step=15 | loss=8.53e-07 | grad_norm=9.87e-05 | rollout=15.9s | train=4.8s
step=20 | loss=8.63e-07 | grad_norm=7.62e-05 | rollout=8.9s  | train=4.9s
```

Loss decreasing, training stable over 20 steps ✓

This was run on the [fork](https://github.com/kahlun/verl) with the full XPU patch stack (device detection + FSDP workers + vLLM rollout fixes). This PR is the device detection foundation — the remaining patches will follow as separate PRs.

**Comment by yao-matrix** [comment]
@wuxibin89 , do you need more information to help your review? Generally, we are trying to enable Intel GPU(device name xpu in PyTorch) support in Verl, we plan to start from `FSDP + vLLM` and expand to more along the way, thx. This first PR is to enable device support.
